### ISSUE TITLE
[Data/Preprocessors]: Preprocessor chain creates multiple dataset pipelines when HASH_SHUFFLE is enabled

### ISSUE THREAD

**Issue by martinbomio** [ray-project/ray#57983]
### What happened + What you expected to happen

It seems like when shuffle_strategy is set to `HASH_SHUFFLE`, multiple dataset pipelines are created, it seems like as many as fittable preprocessors + 1 are created

### Versions / Dependencies

ray 2.49.1

### Reproduction script

```
import ray
import numpy as np
import pandas as pd 
from typing import Dict
from ray.data.preprocessor import Preprocessor
from ray.data.preprocessors import Chain
from ray.data.preprocessors import Normalizer, StandardScaler, SimpleImputer
from ray.data import DatasetContext
from ray.data.context import ShuffleStrategy

ray.init("auto")

DatasetContext.get_current().shuffle_strategy = ShuffleStrategy.HASH_SHUFFLE
ds = ray.data.from_pandas(pd.DataFrame({
    "to_impute": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    "to_normalize": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    "to_scale": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
}))

chain = Chain(
    Normalizer(columns=["to_normalize"]),
    StandardScaler(columns=["to_scale"]),
    SimpleImputer(columns=["to_impute"], strategy="mean")
)
chain.fit_transform(ds).take(10)
```

This creates the 3 dataset pipelines 

<img width="1559" height="859" alt="Image" src="https://github.com/user-attachments/assets/7b4732ee-a935-45d2-8129-41f259783be2" />

If you take a look at each one of them, they are doing one preprocessor at a time, first `Normalizer` (which is not fittabke), then `Normalizer->StandardScaler` and finally `Normalizer->StandardScaler->SimpleImputer`

### Issue Severity

None

**Comment by cem-anyscale** [comment]
@martinbomio this behavior is not specific to `HASH_SHUFFLE` though. Are you comparing the behavior of `AggregateFn` based aggregations to `iter_batches` based computations.
